merkledag/merkledag_cache.go

```diff
@@ -0,0 +1,69 @@
+// package merkledag implements the ipfs Merkle DAG datastructures.
+package merkledag
+
+import (
+	key "github.com/ipfs/go-ipfs/blocks/key"
+	"gx/ipfs/QmVYxfoJQiZijTgPNHCHgHELvQpbsJNTg6Crmc3dQkj3yy/golang-lru"
+	"gx/ipfs/QmZy2y8t9zQH2a1b8q2ZSLKp17ATuJoCNxxyMFG5qFExpt/go-net/context"
+)
+
+func NewDAGCache(ds DAGService, size int) (DAGService, error) {
+	cache, err := lru.New(size)
+	if err != err {
+		return nil, err
+	}
+	return &dagCache{ds, cache}, nil
+}
+
+type dagCache struct {
+	DAGService
+	cache *lru.Cache
+}
+
+func CloneNode(nd * Node) *Node {
```

Comment:
I disagree. We cannot guarantee that callers will not choose to modify nodes they get back from the DAGService. This happens *very frequently*, see the object patch commands, the dagmodifier, the pinset commands, mfs, and a few other places. If a deep copy becomes a perf issue, we can address it then.

> general bad practice to modify an object returned from the cache.

This might be true, but the caller has no idea if the object they are getting was returned from a cache